akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -3264,7 +3264,8 @@ trait FlowOps[+Out, +Mat] {
 
   private def internalConcat[U >: Out, Mat2](that: Graph[SourceShape[U], Mat2], detached: Boolean): Repr[U] =
     that match {
-      case source if source eq Source.empty => this.asInstanceOf[Repr[U]]
+      case source: Source[U @unchecked, _] if source eq Source.empty                 => this.asInstanceOf[Repr[U]]
+      case source: javadsl.Source[U @unchecked, _] if source.asScala eq Source.empty => this.asInstanceOf[Repr[U]]
```

Comment:
I was thinking we could put an internal  `isEmpty(Graph[SourceShape[_], _])` somewhere, like we have `getSingleSource` in `TraversalBuilder` (but this is unrelated to `TraversalBuilder` so somewhere else)

Reply:
I know, but don't find a good place to

Reply:
I just realized that we need to do something just like what getSingleSource does to be sure we don't omit mapMatVal so perhaps it isn't such a bad place after all.